### PR TITLE
sd-device, udev: several cleanups and one udevadm fix

### PR DESCRIPTION
**Pull request by yuwata**
Replaces #18559.

### CODE REVIEW

File: src/libsystemd/sd-device/sd-device.c [hunk 1]
```diff
@@ -1943,29 +1943,14 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
...
-                if (r == -ELOOP)
-                        return -EINVAL;
-                if (r == -EISDIR)
-                        return r;
-
-                r = free_and_strdup(&value, "");
-                if (r < 0)
-                        return r;
-
-                r = device_add_sysattr_value(device, sysattr, value);
```
Comment (keszybz) [inline src/libsystemd/sd-device/sd-device.c:1996]: Hmm, equivalent code was already there when this function was initially added in 57fa1d094cd2c5ac68970526ad0a0754c548e75d. Yeah, it seems better to drop this strange thing.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -1159,56 +1159,54 @@ static int on_ctrl_msg(struct udev_ctrl *uctrl, enum udev_ctrl_msg_type type, co
...
-        filename = strjoina(syspath, "/uevent");
-        log_device_debug(dev, "device is closed, synthesising 'change' on %s", syspath);
-        r = write_string_file(filename, "change", WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (DEBUG_LOGGING) {
+                const char *syspath = NULL;
+                (void) sd_device_get_syspath(target, &syspath);
+                log_device_debug(dev, "device is closed, synthesising 'change' on %s", strna(syspath));
+        }
+
+        r = sd_device_set_sysattr_value(target, "uevent", "change");
```
Comment (keszybz) [inline src/udev/udevd.c]: This feels like a violation of semantics. `sd_device_set_sysattr_value` is for setting of attributes, and caches the attribute. That fact that `uevent` is a bit like an attribute is an accident of implementation. If anything, I think it'd be clearer to add a small helper to write to `uvent`, and not reuse `sd_device_set_sysattr_value` for this.

@poettering ?

File: src/udev/udevadm-trigger.c [hunk 3]
```diff
@@ -41,13 +41,33 @@ static int exec_list(sd_device_enumerator *e, const char *action, Set **settle_s
 
                 r = sd_device_set_sysattr_value(d, "uevent", action);
                 if (r < 0) {
+                        /* ENOENT may be returned when a device does not have /uevent or is already
+                         * removed. Hence, this is logged in debug level and ignored.
```
Comment (keszybz) [inline src/udev/udevadm-trigger.c]: logged at (not in), here and below.

File: src/udev/udevadm-trigger.c [hunk 4]
```diff
@@ -41,13 +41,33 @@ static int exec_list(sd_device_enumerator *e, const char *action, Set **settle_s
...
+                         *
+                         * ENODEV may be returned by some buggy(?) devices e.g. /sys/devices/vio.
+                         * See, https://github.com/systemd/systemd/issues/13652#issuecomment-535129791
+                         * and https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1845319.
+                         * So, this error is also logged in debug level, and ignored.
+                         *
+                         * EROFS is returned when /sys is read only. In that case, all subsequent
+                         * writes should be also failed, hence return earlier.
```
Comment (keszybz) [inline src/udev/udevadm-trigger.c]: should be also failed → will also fail ?
earlier → immediately ?

File: src/udev/udevd.c (hunk 2 above)
Comment (yuwata) [reply src/udev/udevd.c]: Some programs already use `sd_device_set_sysattr_value()` for writing `uevent`. So, if we refuse to use that, then we need to introduce `sd_device_set_action()` or so.

Comment (yuwata) [reply src/udev/udevd.c]: Hmm, https://github.com/opensvc/multipath-tools/blob/66791b63c3ce5feb04cc617c039f060c4edf0988/multipathd/main.c#L1523-L1525

Comment (yuwata) [reply src/udev/udevd.c]: Google says `sd_device_set_sysattr_value()` is not used with `"uevent"`.

**Comment by yuwata** [comment]
@keszybz Thank you for your review.
Updated.
- device_{add,remove}_sysattr_value() are renamed.
- sd_device_get_sysattr_value() refuses to read uevent file.
- sd_device_set_sysattr_value() still accepts to write uevent file, but the value is not cached.
- systemd-udev-trigger.service ignores exit codes of `udevadm trigger`.

File: src/libsystemd/sd-device/sd-device.c [hunk 5]
```diff
@@ -1872,6 +1872,10 @@ _public_ int sd_device_get_sysattr_value(sd_device *device, const char *sysattr,
         assert_return(device, -EINVAL);
         assert_return(sysattr, -EINVAL);
 
+        /* Refuse to read uevent file by using this function. */
+        if (streq(sysattr, "uevent"))
+                return -EINVAL;
```
Comment (keszybz) [inline src/libsystemd/sd-device/sd-device.c]: I don't think we want to break compatibility like that. Even if no publicly visible code does that, sd-device has been around for a while and we don't know all the users.

Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c]: Indeed. Only we need to do is disabling caching when writing uevent files. I have no idea why I decided to add this yesterday :-)

**Comment by yuwata** [comment]
@keszybz Thank you for your comment. Now, `sd_device_get_sysattr_value()` can read uevent file as previous. PTAL.

File: src/libsystemd/sd-device/sd-device.c [hunk 6]
```diff
@@ -1983,29 +1983,14 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
...
-
-                r = device_add_sysattr_value(device, sysattr, value);
-                if (r < 0)
-                        return r;
-                TAKE_PTR(value);
-
-                return -ENXIO;
-        }
+        if (r < 0)
+                return r;
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:1999]: but maybe anything cached here should be dropped in the error path? we don#t really know how this failed after all, and whether the old value is stil in tact

File: src/libsystemd/sd-device/sd-device.c [hunk 7]
```diff
@@ -1968,20 +1967,17 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
...
 
-        len = strlen(_value);
+        value = strdup(_value);
+        if (!value)
+                return -ENOMEM;
 
         /* drop trailing newlines */
-        while (len > 0 && _value[len - 1] == '\n')
-                len --;
+        truncate_nl(value);
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c]: truncate_nl() truncates at first NL, while the above dropped trailing newline.

use instead:

```c
delete_trailing_chars(value, "\n");
```

File: src/libsystemd/sd-device/sd-device.c [hunk 8]
```diff
@@ -1968,20 +1967,17 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
 
         path = prefix_roota(syspath, sysattr);
 
-        len = strlen(_value);
+        value = strdup(_value);
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c]: the old code copied at most 4K, while the new code will copy everything, and then late rcomplain if 4K limit is breached. But I am not sure this matters. this shouldn#t happen IRL, and I like clean easy-to-read code

File: src/libsystemd/sd-device/sd-device.c [hunk 9]
```diff
@@ -1924,8 +1924,7 @@ _public_ int sd_device_get_sysattr_value(sd_device *device, const char *sysattr,
                         return r;
 
                 /* drop trailing newlines */
-                while (size > 0 && value[--size] == '\n')
-                        value[size] = '\0';
+                truncate_nl(value);
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c]: as above

File: src/libsystemd/sd-device/sd-device.c [hunk 10]
```diff
@@ -1894,12 +1894,16 @@ _public_ int sd_device_get_sysattr_value(sd_device *device, const char *sysattr,
...
+
                 /* remember that we could not access the sysattr */
-                r = device_add_sysattr_value(device, sysattr, NULL);
-                if (r < 0)
-                        return r;
+                k = device_add_sysattr_value(device, sysattr, NULL);
+                if (k < 0)
+                        log_device_debug_errno(device, k,
+                                               "sd-device: failed to cache attribute '%s' with NULL, ignoring: %m",
+                                               sysattr);
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:1906]: maybe flush out any old cached value if this fails? or does device_add_sysattr_value() guarantee that anyway?

File: src/udev/udevd.c (hunk 2 above)
Comment (poettering) [reply src/udev/udevd.c]: maybe we should actually have new call sd_device_fire() or so, that fires an uevent off, and encapsulates this? I think that would be pretty nice to have, the file is after all indeed not a regular file but something special.

(or sd_device_emit()? sd_device_uevent()? sd_device_trigger()? sd_device_trigger_uevent()? ideas? – i like shorter better, maybe just sd_device_trigger() given that "udevadm trigger" is ultimately just a big wrapper around this in the end)

File: src/udev/udevadm-trigger.c [hunk 11]
```diff
@@ -41,13 +41,33 @@ static int exec_list(sd_device_enumerator *e, const char *action, Set **settle_s
 
                 r = sd_device_set_sysattr_value(d, "uevent", action);
                 if (r < 0) {
+                        /* ENOENT may be returned when a device does not have /uevent or is already
+                         * removed. Hence, this is logged at debug level and ignored.
+                         *
+                         * ENODEV may be returned by some buggy(?) devices e.g. /sys/devices/vio.
+                         * See, https://github.com/systemd/systemd/issues/13652#issuecomment-535129791
+                         * and https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1845319.
+                         * So, this error is also logged at debug level, and ignored.
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: should be LOG_WARN i think. We want people to fix this, hence the message should be seen

**Review by poettering — CHANGES_REQUESTED** [review]
@brauner, ptal!

**Comment by poettering** [comment]
hmm, one more comment: sd_device_trigger() should take an sd_device_action_t param btw, not a string...

File: src/libsystemd/sd-device/sd-device.c (hunk 10 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:1906]: Yes. `device_add_sysattr_value()` first removes the old entry. So, we do not need to flush on failure.

**Comment by yuwata** [comment]
@poettering Thank you for the review. I force-pushed a revised version. All comments are addressed. PTAL.

File: src/libsystemd/sd-device/sd-device.c (hunk 8 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c]: Indeed, the copied string will be cached. I do not want to cache unnecessarily long string. So, the change is dropped.

File: src/libsystemd/sd-device/sd-device.c [hunk 12]
```diff
@@ -1924,8 +1924,7 @@ _public_ int sd_device_get_sysattr_value(sd_device *device, const char *sysattr,
                         return r;
 
                 /* drop trailing newlines */
-                while (size > 0 && value[--size] == '\n')
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:1927]: given you are not using `size` anymore here, you can remove the variable and pass NULL as last arg to read_full_virtual_file()

**Comment by poettering** [comment]
Looks excellent to me. Would love to hear @brauner's take on this, in particular the last 3 commits in the PR.

File: src/libsystemd/sd-device/sd-device.c (hunk 12 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:1927]: Ouch.

**Comment by yuwata** [comment]
Updated. Upgrading the green label.

**Comment by brauner** [comment]
Testing now.

**Comment by stgraber** [comment]
Did a quick test here and `udevadm trigger` now behaves as expected for me.

**Comment by brauner** [comment]
Fixes the issues for me here. Thank you very much!

**Comment by mrc0mmand** [comment]
TEST-50-DISSECT is still unhappy:

```
[    6.695989] systemd[1]: dev-loop0.device: Collecting.
[    6.696080] testsuite-50.sh[389]: Successfully waited for device 'loop1' to initialize for 504us.
[    6.696153] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=883 reply_cookie=0 signature>
[    6.696229] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=884 reply_cookie=0 signature>
[    6.696311] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=885 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.696377] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    6.696455] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    6.696529] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=888 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.696659] systemd[1]: dev-disk-by\x2dpartlabel-Root\x5cx20Partition.device: Changed plugged -> dead
[    6.696767] systemd[1]: dev-disk-by\x2dpartuuid-0abc16cd\x2da71e\x2db1bd\x2d896c\x2dd9bb45e3dff2.device: Changed plugged -> dead
[    6.696845] systemd[1]: dev-loop0p1.device: Changed plugged -> dead
[    6.696924] systemd[1]: sys-devices-virtual-block-loop0-loop0p1.device: Changed plugged -> dead
[    6.696992] systemd[1]: sys-devices-virtual-block-loop0-loop0p1.device: Collecting.
[    6.697063] systemd[1]: dev-loop0p1.device: Collecting.
[    6.697129] systemd[1]: dev-disk-by\x2dpartuuid-0abc16cd\x2da71e\x2db1bd\x2d896c\x2dd9bb45e3dff2.device: Collecting.
[    6.697194] systemd[1]: dev-disk-by\x2dpartlabel-Root\x5cx20Partition.device: Collecting.
[    6.697259] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartlabel_2dRoot_5cx5cx20Partition_2edevice interface=org.freedesktop.DBus.Properties member=Propertie>
[    6.697333] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartlabel_2dRoot_5cx5cx20Partition_2edevice interface=org.freedesktop.DBus.Properties member=Propertie>
[    6.697405] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=891 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.697470] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartuuid_2d0abc16cd_5cx2da71e_5cx2db1bd_5cx2d896c_5cx2dd9bb45e3dff2_2edevice interface=org.freedesktop>
[    6.697541] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartuuid_2d0abc16cd_5cx2da71e_5cx2db1bd_5cx2d896c_5cx2dd9bb45e3dff2_2edevice interface=org.freedesktop>
[    6.697624] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=894 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.697701] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0p1_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=895 reply_cookie=0 signatu>
[    6.697775] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0p1_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=896 reply_cookie=0 signatu>
[    6.697847] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=897 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.697921] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2dloop0p1_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChang>
[    6.697997] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2dloop0p1_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChang>
[    6.698064] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=900 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.698138] systemd[1]: dev-disk-by\x2duuid-a3ae2509\x2dda27\x2d47bd\x2d83ee\x2d616b21383d7b.device: Changed plugged -> dead
[    6.698216] systemd[1]: dev-disk-by\x2dpartuuid-b97c758d\x2d97e7\x2d3dc3\x2d7188\x2da81ea78d74eb.device: Changed plugged -> dead
[    6.698292] systemd[1]: dev-disk-by\x2dpartlabel-Verity\x5cx20Partition.device: Changed plugged -> dead
[    6.698371] systemd[1]: dev-loop0p2.device: Changed plugged -> dead
[    6.698438] systemd[1]: sys-devices-virtual-block-loop0-loop0p2.device: Changed plugged -> dead
[    6.698508] systemd[1]: sys-devices-virtual-block-loop0-loop0p2.device: Collecting.
[    6.698580] systemd[1]: dev-loop0p2.device: Collecting.
[    6.698664] systemd[1]: dev-disk-by\x2dpartlabel-Verity\x5cx20Partition.device: Collecting.
[    6.698737] systemd[1]: dev-disk-by\x2dpartuuid-b97c758d\x2d97e7\x2d3dc3\x2d7188\x2da81ea78d74eb.device: Collecting.
[    6.698846] systemd[1]: dev-disk-by\x2duuid-a3ae2509\x2dda27\x2d47bd\x2d83ee\x2d616b21383d7b.device: Collecting.
[    6.698915] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2duuid_2da3ae2509_5cx2dda27_5cx2d47bd_5cx2d83ee_5cx2d616b21383d7b_2edevice interface=org.freedesktop.DBu>
[    6.698982] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2duuid_2da3ae2509_5cx2dda27_5cx2d47bd_5cx2d83ee_5cx2d616b21383d7b_2edevice interface=org.freedesktop.DBu>
[    6.699047] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=903 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.699117] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartuuid_2db97c758d_5cx2d97e7_5cx2d3dc3_5cx2d7188_5cx2da81ea78d74eb_2edevice interface=org.freedesktop>
[    6.699199] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartuuid_2db97c758d_5cx2d97e7_5cx2d3dc3_5cx2d7188_5cx2da81ea78d74eb_2edevice interface=org.freedesktop>
[    6.699271] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=906 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.699336] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartlabel_2dVerity_5cx5cx20Partition_2edevice interface=org.freedesktop.DBus.Properties member=Propert>
[    6.699411] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2dpartlabel_2dVerity_5cx5cx20Partition_2edevice interface=org.freedesktop.DBus.Properties member=Propert>
[    6.699494] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=909 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.699565] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0p2_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=910 reply_cookie=0 signatu>
[    6.699648] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0p2_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=911 reply_cookie=0 signatu>
[    6.699719] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=912 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.699782] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2dloop0p2_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChang>
[    6.699853] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2dloop0p2_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChang>
[    6.815055] kernel:  loop1: p1 p2
[    6.700105] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=915 reply_cookie=0 signature=so error-name=n/a error-messag>
[    6.701252] systemd[1]: dev-loop1.device: Changed dead -> plugged
[    6.701363] systemd[1]: sys-devices-virtual-block-loop1.device: Changed dead -> plugged
[    6.701436] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitNew cookie=916 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[    6.701521] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitNew cookie=917 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[    6.702789] testsuite-50.sh[389]: Failed to dissect image '/tmp/tmp.W8g91S/minimal_0.gpt': Operation not permitted
```

Full journal: https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/2984/artifact//systemd-centos-ci/artifacts_7YQVhx/vagrant-logs.LMn/vagrant-arch-testsuite.ZSU/TEST-50-DISSECT/system.journal

**Comment by yuwata** [comment]
`testsuite-50.sh[389]: Failed to dissect image '/tmp/tmp.W8g91S/minimal_0.gpt': Operation not permitted`

@bluca and @mrc0mmand Is this new? Or already detected previously?

**Comment by mrc0mmand** [comment]
> `testsuite-50.sh[389]: Failed to dissect image '/tmp/tmp.W8g91S/minimal_0.gpt': Operation not permitted`
> 
> @bluca and @mrc0mmand Is this new? Or already detected previously?

This is new (and it's not a flake, as it failed in all runs for this PR).

File: src/shared/dissect-image.c [hunk 13]
```diff
@@ -150,7 +150,8 @@ static int device_is_partition(sd_device *d, blkid_partition pp) {
                 return false;
 
         r = sd_device_get_sysattr_value(d, "partition", &v);
-        if (r == -ENOENT) /* Not a partition device */
+        if (r == -ENOENT ||        /* Not a partition device */
+            ERRNO_IS_PRIVILEGE(r)) /* Not ready to access? */
```
Comment (yuwata) [inline src/shared/dissect-image.c:154]: @bluca Could you check this?

**Comment by yuwata** [comment]
- two trivial commits are added at the end.
- hopefully dissect error is fixed, but I am not sure it is a correct way to fix.

PTAL.

File: src/shared/dissect-image.c (hunk 13 above)
Comment (poettering) [reply src/shared/dissect-image.c:154]: Hmm, why would we see EPERM/EACCESS here? the "partition" sysfs attribute is not accessible initially? that's so strange

Comment (yuwata) [reply src/shared/dissect-image.c:154]: Yeah, but it seems `lstat()` returns EPERM.

Comment (poettering) [reply src/shared/dissect-image.c:154]: This is so strange... The old code hid that error code I guess. But I wonder how this happens?

**Comment by yuwata** [comment]
Yey, now all CentOS CIs are green again!

File: src/shared/dissect-image.c (hunk 13 above)
Comment (bluca) [reply src/shared/dissect-image.c:154]: sorry - no clue, it seems very strange

Comment (yuwata) [reply src/shared/dissect-image.c:154]: #18717 hides the error again :-p

File: src/mount/mount-tool.c [hunk 14]
```diff
@@ -1197,7 +1197,7 @@ static int acquire_removable(sd_device *d) {
                 return 0;
 
         for (;;) {
-                if (sd_device_get_sysattr_value(d, "removable", &v) > 0)
+                if (sd_device_get_sysattr_value(d, "removable", &v) >= 0)
```
Comment (keszybz) [inline src/mount/mount-tool.c:1200]: Oops.

**Review by keszybz — COMMENTED** [review]
LGTM.

**Pull request merged**